### PR TITLE
React Native Web: Add framework, CLI integration, sandboxes

### PR DESCRIPTION
**Pull request by shilman**
Closes #

## What I did

Barebones Vite-based React Native Web framework, `@storybook/react-native-web-vite`, based on https://github.com/storybookjs/addon-react-native-web/

- [x] New framework
- [x] Basic CLI integration with `storybook init --type react_native_web`
- [x] Sandboxes for RN CLI and Expo
- [x] CLI templates for example RN `Button`/`Header`/`Page` components/stories
- [x] Figure out JSX settings
- [x] Docgen / auto-generated controls
- [x] Remove CLI-generated `.css` files
- [x] Test sandboxes daily
- [x] Documentation

## Follow-up work
- [ ] Improved CLI integration that prompts user
- [ ] Debug RN-CLI sandbox
- [ ] Set up `react-native-web` renderer docs
- [ ] Debug `storybook add @storybook/experimental-addon-test` (if needed)

#### The changes in this PR are covered in the following automated tests:

- [x] stories
- [ ] unit tests
- [ ] integration tests
- [x] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-29520-sha-c92aa93a`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-29520-sha-c92aa93a sandbox` or in an existing project with `npx storybook@0.0.0-pr-29520-sha-c92aa93a upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-29520-sha-c92aa93a`](https://npmjs.com/package/storybook/v/0.0.0-pr-29520-sha-c92aa93a) |
| **Triggered by** | @shilman |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`shilman/react-native-web-vite`](https://github.com/storybookjs/storybook/tree/shilman/react-native-web-vite) |
| **Commit** | [`c92aa93a`](https://github.com/storybookjs/storybook/commit/c92aa93aa22040d9df0fbb8380d8a12498aec91d) |
| **Datetime** | Sat Nov 16 14:54:44 UTC 2024 (`1731768884`) |
| **Workflow run** | [11871015323](https://github.com/storybookjs/storybook/actions/runs/11871015323) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=29520`_
</details>

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  78.5 MB | 78.5 MB | 0 B | **1.36** | 0% |
| initSize |  144 MB | 144 MB | 791 B | **1.37** | 0% |
| diffSize |  65.2 MB | 65.2 MB | 791 B | **1.58** | 0% |
| buildSize |  6.87 MB | 6.87 MB | 0 B | **-1.36** | 0% |
| buildSbAddonsSize |  1.51 MB | 1.51 MB | 0 B | - | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.9 MB | 1.9 MB | 0 B | 1.18 | 0% |
| buildSbPreviewSize |  271 kB | 271 kB | 0 B | - | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  3.88 MB | 3.88 MB | 0 B | 1.18 | 0% |
| buildPreviewSize |  2.99 MB | 2.99 MB | 0 B | **-1.36** | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.4s | 23.8s | 16.3s | **1.29** | 🔺68.5% |
| generateTime |  22.5s | 20.1s | -2s -410ms | -0.45 | -12% |
| initTime |  14.8s | 15.7s | 946ms | 0.45 | 6% |
| buildTime |  9s | 8s | -952ms | -0.8 | -11.8% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  6.3s | 5.1s | -1s -211ms | -0.84 | -23.4% |
| devManagerResponsive |  4s | 3.2s | -827ms | -0.83 | -25.7% |
| devManagerHeaderVisible |  601ms | 506ms | -95ms | -0.8 | -18.8% |
| devManagerIndexVisible |  637ms | 538ms | -99ms | -1.09 | -18.4% |
| devStoryVisibleUncached |  927ms | 1s | 139ms | 0.36 | 13% |
| devStoryVisible |  631ms | 536ms | -95ms | -1.13 | -17.7% |
| devAutodocsVisible |  606ms | 516ms | -90ms | 0.11 | -17.4% |
| devMDXVisible |  633ms | 439ms | -194ms | -1.06 | -44.2% |
| buildManagerHeaderVisible |  610ms | 501ms | -109ms | -0.91 | -21.8% |
| buildManagerIndexVisible |  624ms | 523ms | -101ms | -0.84 | -19.3% |
| buildStoryVisible |  607ms | 491ms | -116ms | -1.02 | -23.6% |
| buildAutodocsVisible |  510ms | 428ms | -82ms | -0.94 | -19.2% |
| buildMDXVisible |  501ms | 440ms | -61ms | -0.13 | -13.9% |

## Greptile Summary

Added React Native Web Vite framework to Storybook with CLI integration, sandboxes, and test coverage, focusing on proper Vite configuration and React Native Web component support.

- Added `@storybook/react-native-web-vite` framework with Vite-specific configuration in `code/frameworks/react-native-web-vite/src/preset.ts`
- Added CLI templates with TypeScript/JavaScript examples in `code/frameworks/react-native-web-vite/template/cli/`
- Updated framework detection to handle React Native Web projects in `code/core/src/cli/detect.ts` and `code/core/src/cli/project_types.ts`
- Added test skips for React Native Web specific behaviors in e2e tests (onPress vs onClick, CSS differences)
- Added sandbox templates for Expo and RN CLI in `code/lib/cli-storybook/src/sandbox-templates.ts`

### CODE REVIEW

File: docs/get-started/frameworks/react-native-web-vite.mdx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+title: Storybook for React Native Web
+sidebar:
+  order: 5
+  title: React Native Web
+---
+
+Storybook for React Native Web is a [framework](../../contribute/framework.mdx) that makes it easy to develop and test UI components in isolation for [React Native](https://reactnative.dev/) applications. It uses [Vite](https://vitejs.dev/) to build your components for web browsers. The framework includes:
+
+* ⚛️ React Native components
+* 📱 Mobile emulation in the browser
```
Comment (dannyhw) [inline docs/get-started/frameworks/react-native-web-vite.mdx:11]: mobile emulation is not being used so this would be misleading

Comment (dannyhw) [reply docs/get-started/frameworks/react-native-web-vite.mdx:11]: I mean that emulation seems like we're saying a phone runs in the browser

Comment (shilman) [reply docs/get-started/frameworks/react-native-web-vite.mdx]: removed!

**Review by jonniebigodes — APPROVED** [review]

File: code/frameworks/react-native-web-vite/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import type {
+  CompatibleString,
+  StorybookConfig as StorybookConfigBase,
+  TypescriptOptions as TypescriptOptionsBase,
+} from 'storybook/internal/types';
+
+import type { BuilderOptions, StorybookConfigVite } from '@storybook/builder-vite';
+
+import type docgenTypescript from '@joshwooding/vite-plugin-react-docgen-typescript';
```
Comment (valentinpalkovic) [inline code/frameworks/react-native-web-vite/src/types.ts:9]: The `react-vite` framework should export the types for `@joshwooding/vite-plugin-react-docgen-typescript`. Then we can import the types from `@storybook/react-vite` and `react-native-web-vite` does not have to declare it as a dependency, and we make sure that the types are the ones that `react-vite` uses internally.

File: code/frameworks/react-native-web-vite/package.json [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    "*.d.ts",
+    "!src/**/*"
+  ],
+  "scripts": {
+    "check": "jiti ../../../scripts/prepare/check.ts",
+    "prep": "jiti ../../../scripts/prepare/bundle.ts"
+  },
+  "dependencies": {
+    "@joshwooding/vite-plugin-react-docgen-typescript": "0.3.0",
+    "@rollup/pluginutils": "^5.0.2",
```
Comment (valentinpalkovic) [inline code/frameworks/react-native-web-vite/package.json:52]: Can't find any reference to `@rollup/pluginutils`

File: code/frameworks/react-native-web-vite/package.json [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    "@storybook/builder-vite": "workspace:*",
+    "@storybook/react": "workspace:*",
+    "@storybook/react-vite": "workspace:*",
+    "@vitejs/plugin-react": "^4.3.2",
+    "find-up": "^5.0.0",
+    "magic-string": "^0.30.0",
+    "react-docgen": "^7.0.0",
```
Comment (valentinpalkovic) [inline code/frameworks/react-native-web-vite/package.json:59]: I can't find any reference to `react-docgen`

File: code/frameworks/react-native-web-vite/package.json [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    "resolve": "^1.22.8",
+    "tsconfig-paths": "^4.2.0"
```
Comment (valentinpalkovic) [inline code/frameworks/react-native-web-vite/package.json:61]: I can't find any reference to `tsconfig-paths`

File: code/frameworks/react-native-web-vite/package.json [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    "@types/node": "^22.0.0",
+    "typescript": "^5.3.2"
+  },
+  "peerDependencies": {
+    "react": "^16.8.0 || ^17.0.0 || ^18.0.0 || ^19.0.0-beta",
+    "react-dom": "^16.8.0 || ^17.0.0 || ^18.0.0 || ^19.0.0-beta",
+    "react-native": ">=0.74.5",
+    "react-native-web": "^0.19.12",
+    "storybook": "workspace:^",
+    "vite": "^4.0.0 || ^5.0.0"
```
Comment (valentinpalkovic) [inline code/frameworks/react-native-web-vite/package.json:73]: Since this is a new framework, should we be more protective and only allow Vite 5?

Comment (shilman) [reply code/frameworks/react-native-web-vite/package.json]: done

File: code/frameworks/react-native-web-vite/package.json [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  },
+  "engines": {
+    "node": ">=18.0.0"
```
[the remark this answers is not in the document]
Comment (shilman) [reply code/frameworks/react-native-web-vite/package.json:70]: I'm more comfortable making the switch along with all the other frameworks.

**Pull request merged**